torch_xla/csrc/tensor.h

```diff
@@ -426,6 +463,9 @@ class XLATensor {
   static xla::int64 GetNextTensorId();
 
   std::shared_ptr<Data> data_;
+  // For some types (U8, S8 etc), the logical type of the tensor doesn't match
+  // the type of the underlying data.
+  xla::Shape logical_shape_;
```

Comment:
There is no need for a full shape.
This should be at::ScalarType and should be within the Data structure above.

Reply:
What would `shape()` do then, re-create a shape with the changed element type every time we call it? That might be costly.

Reply:
Done.